libexec/chap_md5/chap_md5.8
.\" Copyright (c) 1996 Berkeley Software Design, Inc. All rights reserved.
.\" The Berkeley Software Design Inc. software License Agreement specifies
.\" the terms and conditions for redistribution.
.\"
.\"	BSDI chap_md5.8,v 1.6 1996/11/22 16:17:45 prb Exp
.Dd March 26, 1996
.Dt CHAP_MD5 8
.Os BSD/OS
.Sh NAME
.Nm chap_md5
.Nd preform MD5 CHAP authentication
.Sh SYNOPSIS
.Nm chap_md5
.br
.Nm chap_md5
.Ar user
.Ar id
.br
.Nm chap_md5
.Ar user
.Ar id
.Fl v
.Sh DESCRIPTION
The
.Nm chap_md5
utility is expected to be called by the
.Xr ppp 8
command.
See the
.Li chap-script
option in 
.Xr ppp.sys 5
for a description of the arguments to the
.Nm chap_md5
utility.
.Sh SECRETS
The shared secrets are stored in the file
.Pa /etc/chap_md5_secrets ,
which has lines one of the following formats:
.sp
.Li Ta name:format:secret:
.br
.Li Ta name:challenge:format:secret:
.br
.Li Ta name:response:format:secret:
.sp
.Bl -column XXXXXXX
.It Li name Ta
The user name.
.It Li format Ta
The format the secret is stored in, either
.Li hex
or
.Li ascii
should be used.
.It Li challenge
This string is literal.  This implies the following secret is
only valid for verifying a challenge.
.It Li response
This string is literal.  This implies the following secret is
only valid for generating a response to a challenge.
The
.Li challenge
and
.Li response
keys allow for mutual authentication in which each direction
uses a distinct secret.
.It Li secret Ta
The actual secret.  If the format is
.Li hex
then the secret is an even number of bytes from the set [0-9a-f].
The secret is converted from hex to an actual byte string before use.
If the format is
.Li ascii
then the secret is the actual secret.  Use
.Dq \e\e
to enter a
.Dq \e
and
.Dq \e:
to enter a
.Dq \: .
.Sh FILES
.Pa /etc/chap_md5_secrets
.Sh SEE ALSO
.Xr ppp.sys 5 ,
.Xr ppp 8
.Sh LEGAL
The
.Nm chap_md5
utility uses the RSA Data Security, Inc. MD5 Message-Digest Algorithm.
The following legal notice is required by the md5 source code:
.Bd -literal
License to copy and use this software is granted provided that it
is identified as the "RSA Data Security, Inc. MD5 Message-Digest
Algorithm" in all material mentioning or referencing this software
or this function.

License is also granted to make and use derivative works provided
that such works are identified as "derived from the RSA Data
Security, Inc. MD5 Message-Digest Algorithm" in all material
mentioning or referencing the derived work.
.Ed


libexec/chap_md5/chap_md5.c
/*-
 * Copyright (c) 1996 Berkeley Software Design, Inc.
 * All rights reserved.
 * The Berkeley Software Design Inc. software License Agreement specifies
 * the terms and conditions for redistribution.
 *
 *	BSDI chap_md5.c,v 1.5 1996/11/22 16:17:45 prb Exp
 */

/*
 * Usage: chap_md5			Produce a challenge
 *	  chap_md5 user id		Produce a response
 *	  chap_md5 -v user id		Verify a response
 *
 * When a response is is being generated, the challenge is passed on stdin.
 * When a challenge is being verified, stdin contains the stread:
 * <number-of-bytes-in-response><response...><challenge...>
 * <number-of-bytes-in-response> is a single byte representing between
 * 0 and 255 bytes in the response.  For MD5 this must be 16.
 */


#include <ctype.h>
#include <err.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "md5.h"

#define	MAXMSG		256
#define	MAXSECRET	512
#define	MAXNAME		64

#define	_PATH_MD5_SECRETS	"/etc/chap_md5_secrets"

void getsecret(char *, u_char *, int *, int);

void
main(int ac, char **av)
{
	MD5_CTX md5;

	u_char msg[MAXMSG + MAXSECRET];
	u_char resp[16];
	u_char check[16];
	u_char *m;
	u_char *em = msg + sizeof(msg);
	int vflag = 0;
	int c;
	int i;

	while ((c = getopt(ac, av, "v")) != EOF)
		switch(c) {
		case 'v':
			++vflag;
			break;
		default: usage:
			fprintf(stderr, "Usage: chap_md5 [-v] [user id]\n");
			exit(1);
		}

	if (optind == ac) {
		srandom(getpid() ^ time(0));

		i = (random() & 0x1f) + 0x10;

		for (c = 0; c < i; ++c)
			msg[c] = random() & 0xff;
		if (vflag) {
			for (c = 0; c < i; ++c) {
				printf("%02x", msg[c]);
				if ((c & 0x0f) == 0x0f)
					printf("\n");
			}
			printf("\n");
		} else
			write(1, msg, i);
		exit(0);
	}

	if (optind + 2 != ac)
		goto usage;


	/*
	 * read message in
	 */
	m = msg;
	*m++ = atoi(av[optind+1]);

	getsecret(av[optind], m, &i, vflag);

	m += i;

	if (vflag) {
		char ch;

		if (read(0, &ch, 1) != 1)
			errx(1, "missing response");
		if (ch != 0x10)
			errx(1, "invalid response length"); 
		for (i = 0; i < 0x10; i += c)
			if ((c = read(0, &check[i], 0x10 - i)) <= 0)
				errx(1, "short response"); 
	}

	i = 0;

	while (m < em && (c = read(0, m, em - m)) > 0) {
		i += c;
		m += c;
	}
		
	if (i > MAXMSG)
		errx(1, "message too long");

	MD5Init(&md5);
	MD5Update(&md5, msg, m - msg); 
	MD5Final(resp, &md5);

	if (vflag) {
		if (memcmp(resp, check, 0x10))
			errx(1, "invalid response");
		exit(0);
	}
	write(1, resp, 0x10);
	exit(0);
}

void
getsecret(char *user, u_char *secret, int *lenp, int vflag)
{
	FILE *fp;
	char buf[MAXSECRET + MAXNAME + 16];
	int ulen = strlen(user);
	char *p, *q;

	if ((fp = fopen(_PATH_MD5_SECRETS, "r")) == NULL)
		err(1, _PATH_MD5_SECRETS);
	
	while (fgets(buf, sizeof(buf), fp))
		if (buf[ulen] == ':' && strncmp(user, buf, ulen) == 0) {
			p = buf + ulen;
			if (strncasecmp(p, ":challenge:", 11) == 0) {
				if (!vflag)
					continue;
				p += 10;
			}
			if (strncasecmp(p, ":response:", 10) == 0) {
				if (vflag)
					continue;
				p += 9;
			}
			fclose(fp);
			if ((q = strchr(p+1, ':')) == NULL)
				errx(1, "invalid entry in secret database");
			++q;
			if (strncasecmp(p, ":hex:", 5) == 0) {
				*lenp = 0;
				while (isxdigit(q[0]) && isxdigit(q[1])) {
					*secret++ = (digittoint(q[0]) << 4)
					    | digittoint(q[1]);
					q += 2;
					++*lenp;
				}
				if (*q != ':')
					errx(1, "invalid format in secret database (secret)");
			} else if (strncasecmp(p, ":ascii:", 7) == 0) {
				*lenp = 0;
				while (*q && *q != ':') {
					if (q[0] == '\\' && q[1])
						q++;
					*secret++ = *q++;
					++*lenp;
				}
				if (*q != ':')
					errx(1, "invalid format in secret database (secret)");
			} else
				errx(1, "invalid format in secret database (mode)");
			return;
		}

	fclose(fp);
	errx(1, "%s: user unknown", user);
}
